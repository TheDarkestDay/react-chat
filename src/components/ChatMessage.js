import React from 'react';
import PropTypes from 'prop-types';
import Avatar from 'material-ui/Avatar';
import Paper from 'material-ui/Paper';
import Typography from 'material-ui/Typography';
import { withStyles } from 'material-ui/styles';
import toMaterialStyle from 'material-color-hash';

import generateLastActivityMessage from '../services/generate-last-activity-message';

const messageWrapper = {
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',
  paddingTop: '8px',
  paddingBottom: '8px',
};

const styles = () => ({
  avatar: {
    marginRight: '16px',
  },
  myMessageAvatar: {
    marginLeft: '16px',
    marginRight: 0,
    order: 1,
  },
  message: {
    padding: '8px',
  },
  myMessage: {
    backgroundColor: '#e6dcff',
  },
  messageWrapper,
  myMessageWrapper: {
    ...messageWrapper,
    justifyContent: 'flex-end',
  },
});

const ChatMessage = ({
  classes, sender, content, createdAt, user,
}) => {
  const materialStyles = toMaterialStyle(sender.username);

  return (
    /* eslint-disable no-underscore-dangle */
    <div className={user._id === sender._id ? classes.myMessageWrapper : classes.messageWrapper}>
      <Avatar
        className={user._id === sender._id ? classes.myMessageAvatar : classes.avatar}
        style={materialStyles}
      >
        {sender.username[0]}
      </Avatar>
      <Paper className={`${classes.message} ${user._id === sender._id ? classes.myMessage : ''}`}>
        <Typography variant="caption" style={{ color: materialStyles.backgroundColor }}>
          {sender.username}
        </Typography>
        <Typography>
          {content}
        </Typography>
        <Typography variant="caption">
          {generateLastActivityMessage(createdAt)}
        </Typography>
      </Paper>
    </div>
  );
};

ChatMessage.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  sender: PropTypes.shape({
    username: PropTypes.string.isRequired,
    firstName: PropTypes.string,
    lastName: PropTypes.string,
  }).isRequired,
  content: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
    firstName: PropTypes.string,
    lastName: PropTypes.string,
  }).isRequired,
};

export default withStyles(styles)(ChatMessage);
