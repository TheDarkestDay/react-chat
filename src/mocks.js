/* eslint-disable import/prefer-default-export */
export const userMock = {
  _id: '1',
  username: 'test_user',
};
