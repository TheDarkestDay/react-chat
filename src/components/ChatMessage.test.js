/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import ChatMessage from './ChatMessage';

import { userMock } from '../mocks';

jest.mock('../services/generate-last-activity-message', () => () => '2 days ago');

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <ChatMessage
      sender={userMock}
      user={userMock}
      content="Hello, world!"
      createdAt="01-01-1990"
    />,
    div,
  );
  ReactDOM.unmountComponentAtNode(div);
});

it('should match snapshot', () => {
  const tree = renderer
    .create(<ChatMessage
      sender={userMock}
      user={userMock}
      content="Hello, world!"
      createdAt="01-01-1990"
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
