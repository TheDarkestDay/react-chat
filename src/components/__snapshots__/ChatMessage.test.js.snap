// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="ChatMessage-myMessageWrapper-66"
>
  <div
    className="MuiAvatar-root-67 MuiAvatar-colorDefault-68 ChatMessage-myMessageAvatar-62"
    style={
      Object {
        "backgroundColor": "#795548",
        "color": "rgba(255, 255, 255, 1)",
        "materialColorName": "Brown",
      }
    }
  >
    t
  </div>
  <div
    className="MuiPaper-root-70 MuiPaper-shadow2-74 MuiPaper-rounded-71 ChatMessage-message-63 ChatMessage-myMessage-64"
  >
    <span
      className="MuiTypography-root-97 MuiTypography-caption-107"
      style={
        Object {
          "color": "#795548",
        }
      }
    >
      test_user
    </span>
    <p
      className="MuiTypography-root-97 MuiTypography-body1-106"
    >
      Hello, world!
    </p>
    <span
      className="MuiTypography-root-97 MuiTypography-caption-107"
    >
      2 days ago
    </span>
  </div>
</div>
`;
